macros/src/lib.rs

```diff
@@ -429,7 +429,11 @@ fn fields(
                         "?".to_string()
                     });
                     if let Some(ident) = f.ident.as_ref() {
-                        core::write!(format, "{}: {{={}}}", ident, ty).ok();
+                        if ty == "str" {
```

Comment:
the problem with this approach (attaching `:?` only to the `=str` type) is that it won't work with generic structures. For example:
``` rust
#[derive(Format)]
struct S1<T> {
    x: T,
    y: u8,
}

// outputs: "S { x: hi, y: 42 }"
defmt::info!("{}", S1 { x: "hi", y: 42 });
// outputs: "S { x: hi, y: 0x2a }"
defmt::info!("{:x}", S1 { x: "hi", y: 42 });
```

This is because the expansion of the macro will use the format string `"S1 {{ x: {=?}, y: {=u8} }}"`. The proc-macro cannot know whether the generic type `T` will be instantiated to `str` or not. Also, you don't the `:?` format specifier if `T` is instantiated to a type that's not a string.

Reply:
Thanks for the feedback. The code is changed now to apply the Debug hint `:?` to all inner types. As a side effect this changes the output in the `log` test program as well.

```rust
#[derive(Format)]
struct Data<'a> {
    name: &'a [u8],
    value: bool,
}

let data = &[Data { name: b"Hi", value: true }];
defmt::info!("{=[?]:a}", *data);

// before the change
// -> INFO [Data { name: b"Hi", value: true }]
// after the change
// -> INFO [Data { name: [72, 105], value: true }]
```